qiskit/transpiler/passes/optimization/inverse_cancellation.py

```diff
@@ -0,0 +1,134 @@
+# that they have been altered from the originals.
+
+"""
+A generic gate-inverse cancellation pass for any set of gate-inverse pairs.
+"""
+from typing import List, Tuple, Union
+
+from qiskit.circuit import Gate
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.exceptions import TranspilerError
+
+
+class Cancellation(TransformationPass):
+    """Cancel specific Gates which are inverses of each other when they occur back-to-
+    back."""
+
+    def __init__(self, gates_to_cancel: List[Union[Gate, Tuple[Gate, Gate]]]):
+        """Initialize Cancellation pass.
+
+        Args:
+            gates_to_cancel: list of gates to cancel
+
+        Raises:
+            TranspilerError:
+                Initalization raises an error when the input is not a self-inverse gate
+                or a two-tuple of inverse gates.
+        """
+        for gates in gates_to_cancel:
+            if isinstance(gates, Gate):
+                if gates != gates.inverse():
+                    raise TranspilerError("Gate {} is not self-inverse".format(gates.name))
+            elif isinstance(gates, tuple):
+                if len(gates) != 2:
+                    raise TranspilerError(
+                        "Too many or too few inputs: {}. Only two are allowed.".format(gates)
+                    )
+                elif gates[0] != gates[1].inverse():
+                    raise TranspilerError(
+                        "Gate {} and {} are not inverse.".format(gates[0].name, gates[1].name)
+                    )
+            else:
+                raise TranspilerError(
+                    "Cancellation pass does not take input type {}. Input must be"
+                    " a Gate.".format(type(gates))
+                )
+
+        self.gates_to_cancel = gates_to_cancel
+        super().__init__()
+
+    def run(self, dag: DAGCircuit):
+        """Run the Cancellation pass on `dag`.
+
+        Args:
+            dag: the directed acyclic graph to run on.
+
+        Returns:
+            DAGCircuit: Transformed DAG.
+        """
+        self_inverse_gates = []
+        inverse_gate_pairs = []
+
+        for gates in self.gates_to_cancel:
+            if isinstance(gates, Gate):
+                self_inverse_gates.append(gates)
+            else:
+                inverse_gate_pairs.append(gates)
+
+        dag = self._run_on_self_inverse(dag, self_inverse_gates)
+        return self._run_on_inverse_pairs(dag, inverse_gate_pairs)
+
+    def _run_on_self_inverse(self, dag: DAGCircuit, self_inverse_gates: List[Gate]):
+        """
+        Run self-inverse gates on `dag`.
+
+        Args:
+            dag: the directed acyclic graph to run on.
+            self_inverse_gates: list of gates who cancel themeselves in pairs
+
+        Returns:
+            DAGCircuit: Transformed DAG.
+        """
+        gate_cancel_runs = dag.collect_runs([gate.name for gate in self_inverse_gates])
```

Comment:
```suggestion
        gate_cancel_runs = [dag.collect_runs([gate.name]) for gate in self_inverse_gates]
```
I think there's a small bug here in that `dag.collect_runs` will build runs of gates which match any name in the list instead of making a separate run for each name in the list, which I think is what you want here.

The test `test_inverse_cancellation_hcx` comes close to triggering this, but since `h` and `cx` have a different number of `qargs`, they're split into different `partitions` below. Maybe try a test case involving e.g. `h` and `x` gates?